File: sdk/ts/src/events.ts
import { SDKError } from "./verifier.js";

export const PERMISSION_EVENT_SUBJECTS = [
  "iam.perm.changed",
  "iam.user.disabled",
  "iam.role.updated",
] as const;

export const PermissionEventSubjects = PERMISSION_EVENT_SUBJECTS;

export interface PermissionEventSource {
  subscribe(
    subject: string,
    handler: (message: unknown) => void | Promise<void>,
  ): unknown;
  close?(): void | Promise<void>;
  drain?(): void | Promise<void>;
}

interface PermissionCache {
  invalidate(...userIDs: string[]): void;
}

/** Raised only when a non-empty NATS subscription is requested without nats installed. */
export class NATSDependencyError extends SDKError {
  public constructor(cause?: unknown) {
    super("NATS_UNAVAILABLE", "NATS support requires the optional `nats` peer dependency", cause);
  }
}

/** Handle returned by SubscribePermissions. */
export class PermissionSubscription {
  private closed = false;
  private readonly closer: () => void | Promise<void>;

  public constructor(closer: () => void | Promise<void>) {
    this.closer = closer;
  }

  public async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    await this.closer();
  }

  public async unsubscribe(): Promise<void> {
    await this.close();
  }

  public async Unsubscribe(): Promise<void> {
    await this.close();
  }

  public async Close(): Promise<void> {
    await this.close();
  }
}

export { NATSDependencyError as NATSUnavailableError };
export { NATSDependencyError as NATSError };

/** Subscribe a permissions cache to NATS or a framework-neutral event source. */
export async function subscribePermissions(
  client: PermissionCache | null | undefined,
  sourceOrURL: unknown,
  handler?: (userIDs: readonly string[]) => void | Promise<void>,
): Promise<PermissionSubscription | null> {
  if (client === null || client === undefined) {
    throw new Error("nil permissions client");
  }
  if (typeof sourceOrURL === "string") {
    if (sourceOrURL.trim() === "") return null;
    const source = await connectNATS(sourceOrURL.trim());
    return subscribeSource(client, source, handler);
  }
  if (sourceOrURL === null || sourceOrURL === undefined) return null;
  if (!isPermissionEventSource(sourceOrURL)) {
    throw new TypeError("permission subscription source must provide subscribe(subject, handler)");
  }
  return subscribeSource(client, sourceOrURL, handler);
}

function subscribeSource(
  client: PermissionCache,
  source: PermissionEventSource,
  handler?: (userIDs: readonly string[]) => void | Promise<void>,
): PermissionSubscription {
  const subscriptions: Array<{ unsubscribe?: () => void | Promise<void> }> = [];
  const onMessage = async (message: unknown): Promise<void> => {
    const userIDs = extractUserIDs(message);
    if (userIDs.length === 0) return;
    client.invalidate(...userIDs);
    if (handler !== undefined) await handler(userIDs);
  };
  for (const subject of PERMISSION_EVENT_SUBJECTS) {
    const result = source.subscribe(subject, onMessage);
    if (isPromiseLike(result)) {
      void Promise.resolve(result).then((resolved) => {
        if (isUnsubscriber(resolved)) subscriptions.push(resolved);
      }).catch(() => undefined);
    } else if (isUnsubscriber(result)) {
      subscriptions.push(result);
    }
  }
  return new PermissionSubscription(async () => {
    let firstError: unknown;
    for (const subscription of subscriptions) {
      try {
        await subscription.unsubscribe?.();
      } catch (error) {
        firstError ??= error;
      }
    }
    try {
      await source.drain?.();
      await source.close?.();
    } catch (error) {
      firstError ??= error;
    }
    if (firstError !== undefined) throw firstError;
  });
}

async function connectNATS(url: string): Promise<PermissionEventSource> {
  let moduleValue: unknown;
  try {
    // The optional peer must not be imported while the core SDK is loaded.
    const load = Function("specifier", "return import(specifier)") as (specifier: string) => Promise<unknown>;
    moduleValue = await load("nats");
  } catch (error) {
    throw new NATSDependencyError(error);
  }
  if (!isObject(moduleValue) || !("connect" in moduleValue) || typeof moduleValue.connect !== "function") {
    throw new NATSDependencyError(new Error("nats.connect is unavailable"));
  }
  let connection: unknown;
  try {
    connection = await moduleValue.connect(url);
  } catch (error) {
    throw new NATSDependencyError(error);
  }
  if (!isNATSConnection(connection)) {
    throw new NATSDependencyError(new Error("nats connection does not provide subscribe"));
  }
  return {
    subscribe(subject, handler) {
      const subscription = connection.subscribe(subject);
      void consumeNATS(subscription, handler);
      return {
        unsubscribe: async () => {
          await subscription.unsubscribe?.();
        },
      };
    },
    drain: () => connection.drain?.(),
    close: () => connection.close?.(),
  };
}

async function consumeNATS(
  subscription: AsyncIterable<unknown>,
  handler: (message: unknown) => void | Promise<void>,
): Promise<void> {
  try {
    for await (const message of subscription) {
      await handler(message);
    }
  } catch {
    // A closed subscription ends its iterator; event delivery is best effort.
  }
}

function extractUserIDs(message: unknown): string[] {
  const payload = decodeMessage(message);
  if (!isObject(payload)) return [];
  const IDs: string[] = [];
  const arrayValue = payload.user_ids;
  if (Array.isArray(arrayValue)) {
    for (const value of arrayValue) {
      if (typeof value === "string" && value.trim() !== "") IDs.push(value.trim());
    }
  }
  if (typeof payload.user_id === "string" && payload.user_id.trim() !== "") {
    IDs.push(payload.user_id.trim());
  }
  return [...new Set(IDs)];
}

function decodeMessage(message: unknown): unknown {
  if (typeof message === "string") {
    try {
      return JSON.parse(message);
    } catch {
      return undefined;
    }
  }
  if (message instanceof Uint8Array) {
    try {
      return JSON.parse(new TextDecoder().decode(message));
    } catch {
      return undefined;
    }
  }
  if (isObject(message) && "data" in message) {
    return decodeMessage(message.data);
  }
  return message;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function isPermissionEventSource(value: unknown): value is PermissionEventSource {
  return isObject(value) && "subscribe" in value && typeof value.subscribe === "function";
}

function isUnsubscriber(value: unknown): value is { unsubscribe: () => void | Promise<void> } {
  return isObject(value) && "unsubscribe" in value && typeof value.unsubscribe === "function";
}

function isPromiseLike(value: unknown): value is PromiseLike<unknown> {
  return isObject(value) && "then" in value && typeof value.then === "function";
}

function isNATSConnection(value: unknown): value is {
  subscribe(subject: string): AsyncIterable<unknown> & { unsubscribe?: () => void | Promise<void> };
  drain?(): void | Promise<void>;
  close?(): void | Promise<void>;
} {
  return isObject(value) && "subscribe" in value && typeof value.subscribe === "function";
}

File: sdk/python/src/teamusers_sdk/events.py
"""Optional permission-cache invalidation events."""

from __future__ import annotations

import asyncio
import inspect
import json
import threading
from collections.abc import Callable, Mapping, Sequence
from typing import Any

from .verifier import SDKError

PERMISSION_EVENT_SUBJECTS = (
    "iam.perm.changed",
    "iam.user.disabled",
    "iam.role.updated",
)


class NATSSubscriptionError(SDKError):
    """NATS could not be loaded or a subscription could not be created."""

    code = "NATS_SUBSCRIPTION_FAILED"


class NATSUnavailableError(NATSSubscriptionError):
    """The optional nats-py package is not installed."""

    code = "NATS_DEPENDENCY_MISSING"


class PermissionSubscription:
    """A closeable permission event subscription."""

    def __init__(self, closers: Sequence[Callable[[], Any]] = ()) -> None:
        self._closers = list(closers)
        self._lock = threading.Lock()
        self._closed = False

    def close(self) -> None:
        with self._lock:
            if self._closed:
                return
            self._closed = True
            closers = list(self._closers)
            self._closers.clear()
        first_error: BaseException | None = None
        for closer in closers:
            try:
                result = closer()
                if inspect.isawaitable(result):
                    _run_awaitable(result)
            except BaseException as error:
                if first_error is None:
                    first_error = error
        if first_error is not None:
            raise first_error

    def Close(self) -> None:
        self.close()


# A short alias is convenient in type annotations and mirrors the Go name.
Subscription = PermissionSubscription


def subscribe_permissions(
    client: Any,
    source_or_url: Any,
    handler: Callable[[list[str]], Any] | None = None,
) -> PermissionSubscription | None:
    """Subscribe to the three invalidation subjects.

    ``source_or_url`` may be a NATS URL or an injected synchronous source.  The
    source seam keeps tests and applications that already own a NATS connection
    independent of the optional nats-py package.
    """

    if client is None:
        raise NATSSubscriptionError("permission client is unavailable")
    if isinstance(source_or_url, str):
        if not source_or_url.strip():
            return None
        try:
            import nats  # type: ignore[import-not-found]
        except ImportError as error:
            raise NATSUnavailableError(
                "NATS support requires the optional 'nats-py' extra; install teamusers-sdk[nats]"
            ) from error
        source = _NATSSource(nats, source_or_url.strip())
    else:
        if source_or_url is None:
            raise NATSSubscriptionError("NATS subscription source is required")
        source = source_or_url

    callback = _event_callback(client, handler)
    try:
        return _subscribe_source(source, callback)
    except NATSSubscriptionError:
        raise
    except Exception as error:
        raise NATSSubscriptionError("subscribe to permission events", error) from error


def SubscribePermissions(
    client: Any,
    source_or_url: Any,
    handler: Callable[[list[str]], Any] | None = None,
) -> PermissionSubscription | None:
    return subscribe_permissions(client, source_or_url, handler)


def _event_callback(client: Any, handler: Callable[[list[str]], Any] | None):
    def on_message(*message: Any) -> None:
        payload = message[-1] if message else None
        event = _decode_event(payload)
        if event is None:
            return
        values = event.get("user_ids")
        if values is None:
            values = event.get("userIDs")
        if values is None and isinstance(event.get("user_id"), str):
            values = [event["user_id"]]
        if not isinstance(values, Sequence) or isinstance(values, (str, bytes, bytearray)):
            return
        user_ids = [value for value in values if isinstance(value, str) and value.strip()]
        if not user_ids:
            return
        client.Invalidate(*user_ids)
        if handler is not None:
            handler(list(user_ids))

    return on_message


def _decode_event(payload: Any) -> Mapping[str, Any] | None:
    if hasattr(payload, "data"):
        payload = payload.data
    if isinstance(payload, Mapping):
        return payload
    if isinstance(payload, bytes):
        try:
            payload = payload.decode("utf-8")
        except UnicodeDecodeError:
            return None
    if isinstance(payload, str):
        try:
            value = json.loads(payload)
        except (TypeError, ValueError):
            return None
        return value if isinstance(value, Mapping) else None
    return None


def _subscribe_source(source: Any, callback: Callable[..., Any]) -> PermissionSubscription:
    closers: list[Callable[[], Any]] = []
    if isinstance(source, _NATSSource):
        return source.subscribe(callback)

    subscribe = getattr(source, "subscribe", None)
    if callable(subscribe):
        for subject in PERMISSION_EVENT_SUBJECTS:
            result = _invoke_subscribe(subscribe, subject, callback)
            closer = _closer_for(result)
            if closer is not None:
                closers.append(closer)
        source_closer = _closer_for(source)
        if source_closer is not None:
            closers.append(source_closer)
        return PermissionSubscription(closers)

    if callable(source):
        try:
            result = _invoke_factory(source, callback)
        except Exception as error:
            raise NATSSubscriptionError("create NATS subscription", error) from error
        if isinstance(result, PermissionSubscription):
            return result
        closer = _closer_for(result)
        if closer is not None:
            closers.append(closer)
        return PermissionSubscription(closers)

    raise NATSSubscriptionError("subscription source must provide subscribe() or be callable")


def _invoke_subscribe(subscribe: Callable[..., Any], subject: str, callback: Callable[..., Any]) -> Any:
    try:
        signature = inspect.signature(subscribe)
        positional = [
            parameter
            for parameter in signature.parameters.values()
            if parameter.kind in (parameter.POSITIONAL_ONLY, parameter.POSITIONAL_OR_KEYWORD)
        ]
        if any(parameter.kind == parameter.VAR_POSITIONAL for parameter in signature.parameters.values()) or len(positional) >= 2:
            return subscribe(subject, callback)
        return subscribe(subject=subject, callback=callback)
    except (TypeError, ValueError):
        return subscribe(subject, callback)


def _invoke_factory(factory: Callable[..., Any], callback: Callable[..., Any]) -> Any:
    try:
        signature = inspect.signature(factory)
        positional = [
            parameter
            for parameter in signature.parameters.values()
            if parameter.kind in (parameter.POSITIONAL_ONLY, parameter.POSITIONAL_OR_KEYWORD)
        ]
        if any(parameter.kind == parameter.VAR_POSITIONAL for parameter in signature.parameters.values()) or len(positional) >= 2:
            return factory(PERMISSION_EVENT_SUBJECTS, callback)
        return factory(callback)
    except (TypeError, ValueError):
        return factory(PERMISSION_EVENT_SUBJECTS, callback)


def _closer_for(value: Any) -> Callable[[], Any] | None:
    if value is None:
        return None
    if callable(value):
        return value
    for name in ("unsubscribe", "close", "stop", "Close"):
        method = getattr(value, name, None)
        if callable(method):
            return method
    return None


def _run_awaitable(value: Any) -> Any:
    try:
        asyncio.get_running_loop()
    except RuntimeError:
        return asyncio.run(value)
    result: list[Any] = []
    error: list[BaseException] = []

    def runner() -> None:
        try:
            result.append(asyncio.run(value))
        except BaseException as exc:
            error.append(exc)

    thread = threading.Thread(target=runner, daemon=True)
    thread.start()
    thread.join()
    if error:
        raise error[0]
    return result[0] if result else None


class _NATSSource:
    """Small sync bridge around nats-py's asyncio client."""

    def __init__(self, nats_module: Any, url: str) -> None:
        self._nats = nats_module
        self._url = url

    def subscribe(self, callback: Callable[..., Any]) -> PermissionSubscription:
        ready = threading.Event()
        state: dict[str, Any] = {}
        stopped = threading.Event()

        def runner() -> None:
            async def run() -> None:
                try:
                    connection = await self._nats.connect(self._url)
                    state["connection"] = connection
                    for subject in PERMISSION_EVENT_SUBJECTS:
                        async def on_message(message: Any, _callback: Callable[..., Any] = callback) -> None:
                            _callback(message)

                        await connection.subscribe(subject, cb=on_message)
                    ready.set()
                    while not stopped.is_set():
                        await asyncio.sleep(0.05)
                    await connection.drain()
                except BaseException as error:
                    state["error"] = error
                    ready.set()

            asyncio.run(run())

        thread = threading.Thread(target=runner, name="teamusers-nats", daemon=True)
        thread.start()
        if not ready.wait(10):
            stopped.set()
            raise NATSSubscriptionError("timed out connecting to NATS")
        if "error" in state:
            stopped.set()
            raise NATSSubscriptionError("connect to NATS", state["error"])

        def close() -> None:
            stopped.set()
            thread.join(timeout=10)

        return PermissionSubscription([close])


__all__ = [
    "NATSSubscriptionError",
    "NATSUnavailableError",
    "PERMISSION_EVENT_SUBJECTS",
    "PermissionSubscription",
    "SubscribePermissions",
    "Subscription",
    "subscribe_permissions",
]
